feat: add check for enforcing jwt and lms user email match

syedsajjadkazmii opened the pull request:
**Description:**

Unauthenticate in case of mismatch between JWT email and lms user email

**JIRA:**

[VAN-1694](https://openedx.atlassian.net/browse/VAN-1694)

robrap commented on edx_rest_framework_extensions/auth/jwt/authentication.py:
```diff
@@ -106,6 +113,20 @@ def authenticate(self, request):
                 set_custom_attribute('jwt_auth_result', 'n/a')
                 return user_and_auth
 
+            lms_user_id_property_name = get_setting(VERIFY_LMS_USER_ID_PROPERTY_NAME)
+            # Check lms_user_id_property_name to be 'id' to make this check work for LMS only
+            if lms_user_id_property_name == 'id':
+                user, auth = user_and_auth
+                is_email_mismatch = self._is_jwt_and_lms_user_email_mismatch(request, user)
+                if is_email_mismatch and get_setting(ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH):
+                    set_custom_attribute('jwt_auth_result', 'user-email-mismatch-failure')
+                    raise JwtUserEmailMismatchError(
+                        'Failing JWT authentication due to jwt user email mismatch '
+                        'with lms user email.'
+                    )
+                elif is_email_mismatch:
+                    set_custom_attribute('jwt_auth_result', 'user-email-mismatch-failure')
```
Two changes:
1. I added a new and separate custom attribute `jwt_email_mismatch`.
2. I removed the `elif`, because `jwt_auth_result` will be overridden lower down based on the actual result.
3. Nit: To keep this method cleaner, I'd move the `VERIFY_LMS_USER_ID_PROPERTY_NAME` check inside `_is_jwt_and_lms_user_email_mismatch`, and just return early. 
```suggestion
            is_email_mismatch = self._is_jwt_and_lms_user_email_mismatch(request, user_and_auth[0])
            if is_email_mismatch:
                # .. custom_attribute_name: jwt_email_mismatch
                # .. custom_attribute_description: Temporary attribute to show if email mismatched
                #     even when the enforce toggle is disabled.
                set_custom_attribute('jwt_email_mismatch', is_email_mismatch)
                if get_setting(ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH):
                    set_custom_attribute('jwt_auth_result', 'user-email-mismatch-failure')
                    raise JwtUserEmailMismatchError(
                        'Failing JWT authentication due to jwt user email mismatch '
                        'with lms user email.'
                    )
```

robrap replied:
@syedsajjadkazmii: Ignore the code change I made above. I've changed my mind on some things:
1. Do not use `VERIFY_LMS_USER_ID_PROPERTY_NAME`. It may go away.
2. `ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH` will need to be a permanent toggle, and should be documented as such. Although it will be a temporary rollout toggle for the LMS, ultimately it will be a permanent toggle for the LMS (as the identity server), and will be disabled for all other services.
3. To simplify the code, since this will be permanently disabled elsewhere, you can drop the `jwt_email_mismatch` custom attribute. This means you won't be able to get observability in the LMS before enabling, but hopefully previous testing will be enough.

So code might just look like:
```
if get_setting(ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH):
    is_email_mismatch = self._is_jwt_and_lms_user_email_mismatch(request, user_and_auth[0])
    if is_email_mismatch:
        ...
```

syedsajjadkazmii replied:
Hi @robrap,

Thanks for the suggestions. I have updated the PR with the latest comments. Please have a look.

robrap commented on edx_rest_framework_extensions/auth/jwt/authentication.py:
```diff
@@ -305,6 +326,27 @@ def _is_jwt_cookie_and_session_user_mismatch(self, request):
+        decoded_jwt = configured_jwt_decode_handler(token)
+        jwt_user_email = decoded_jwt.get('email', None)
+
+        if lms_user_email != jwt_user_email:
```
Nit: You could just return the result, but this is up to you.
```suggestion
        return lms_user_email != jwt_user_email:
```

robrap reviewed (commented):
Looks good. Minor issue with the custom attribute. I know you plan to add tests. Thanks and good luck.

robrap reviewed (commented):
Generally looks good. Thanks.
1. I'm still reviewing tests.
2. The changelog and version need to be updated.
3. The base branch will be updated over the next couple of days. Sorry about that. :)

syedsajjadkazmii commented:
> 2. The changelog and version need to be updated.

I will update the changelog and version once my branch is rebased on the master. If it's okay.

robrap commented on edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py:
```diff
@@ -751,6 +757,121 @@ def test_authenticate_jwt_and_no_session_and_set_request_user(self, mock_set_cus
         assert response.status_code == 200
 
+    @override_settings(
+        EDX_DRF_EXTENSIONS={
+            ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH: False
+        },
```
[nit] It might make sense to drop this so we can also test that False is the default.

robrap commented on edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py:
```diff
@@ -751,6 +757,121 @@ def test_authenticate_jwt_and_no_session_and_set_request_user(self, mock_set_cus
+        ROOT_URLCONF='edx_rest_framework_extensions.auth.jwt.tests.test_authentication',
+    )
+    @mock.patch('edx_rest_framework_extensions.auth.jwt.authentication.set_custom_attribute')
+    def test_authenticate_user_lms_and_jwt_email_mismatch_success(self, mock_set_custom_attribute):
```
[nit] Renaming to `test_authenticate_user_lms_and_jwt_email_match_success` would be slightly more clear.

robrap commented on edx_rest_framework_extensions/config.py:
```diff
@@ -17,6 +17,16 @@
+# .. toggle_implementation: DjangoSetting
+# .. toggle_default: False
+# .. toggle_description: Toggle to add a check for matching user email in JWT and LMS user email
+#       for authentication in JwtAuthentication class.
```
Could you add something like the following sentence?
```
This toggle should only be enabled in the LMS as our identity service that is also creating the JWTs.
```

robrap reviewed (commented):
Thanks.

syedsajjadkazmii commented:
Hi @robrap, 

Could you please have a final look at the PR if possible? Also, I was wondering when will your [branch](https://github.com/openedx/edx-drf-extensions/pull/404) get merged into the master?

robrap reviewed (commented):
Thanks for all the work. 
1. My change was merged, so you can rebase against master.
2. I just looked at my final review comments and marked resolved. I did not re-review all the tests, etc.
3. I give this a 👍, except it is difficult to review until you rebase. Feel free to have someone on your team do the final review after rebase (mostly to make sure the rebase looks good), or I can do that if you need me to.

syedsajjadkazmii commented:
> Thanks for all the work.
> 
> 1. My change was merged, so you can rebase against master.
> 2. I just looked at my final review comments and marked resolved. I did not re-review all the tests, etc.
> 3. I give this a 👍, except it is difficult to review until you rebase. Feel free to have someone on your team do the final review after rebase (mostly to make sure the rebase looks good), or I can do that if you need me to.
> 4. Remember to add CHANGELOG and version update for 10.1.0.

Thank you for merging your changes. I have rebased the branch against the master and added CHANGELOG and version updates. Could you please review the tests now?

robrap commented on CHANGELOG.rst:
```diff
@@ -12,6 +12,14 @@ Change Log
+[10.1.0] - 2024-01-26
+---------------------
+
+* Added toggle EDX_DRF_EXTENSIONS[ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH]:
+
+    * If enabled, makes sure that the user email in JWT cookies and LMS user email matches
+    * If email matches, it allows authentication otherwise raise JwtUserEmailMismatchError error.
```
```suggestion
* Added permanent toggle EDX_DRF_EXTENSIONS[ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH]:

    * This toggle should only get enabled in the LMS, and should remain disabled in all other services.
    * If enabled, makes sure that the user email in JWT cookies and LMS user email matches
    * If email matches, it allows authentication otherwise raise JwtUserEmailMismatchError error.
```

robrap commented on edx_rest_framework_extensions/auth/jwt/tests/test_authentication.py:
```diff
@@ -534,6 +537,117 @@ def test_authenticate_jwt_and_no_session_and_set_request_user(self, mock_set_cus
+
+        assert response.status_code == 200
+
+    @override_settings(
```
Note: I just read about `modify_settings`. I'm not sure if you can use both, but for `EDX_DRF_EXTENSIONS`, you might be able to make the test more clear by just updating `ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH`?

syedsajjadkazmii replied:
Unfortunately, `modify_settings` doesn't seem to work with the nested settings in our case. 

> but for EDX_DRF_EXTENSIONS, you might be able to make the test more clear by just updating `ENABLE_JWT_AND_LMS_USER_EMAIL_MATCH`?

if we don't override `JWT_PAYLOAD_USER_ATTRIBUTE_MAPPING: {},` then it gets these [default values](https://github.com/openedx/edx-drf-extensions/blob/master/edx_rest_framework_extensions/settings.py#L29-L32) containing `email`. and on mismatch between the JWT and LMS user email, it simply gets updated at this part [here](https://github.com/openedx/edx-drf-extensions/blob/master/edx_rest_framework_extensions/auth/jwt/authentication.py#L206-L209). As a result, the emails get matched in our condition in `_is_jwt_and_lms_user_email_mismatch` function thus the `JwtAuthentication` returns 200 and does not fail. 

As LMS [uses](https://github.com/edx/edx-internal/blob/1301473ad43e1147d8a046056fec02e533abdb67/edx-remote-config/stage/lms.yml#L875) `JWT_PAYLOAD_USER_ATTRIBUTE_MAPPING: {},`, so I have to override it in the tests.

robrap reviewed (approved):
Minor request to update the changelog. Thank you.
